posts/java-enums-with-extra-visitor-please.md

```diff
@@ -0,0 +1,203 @@
+
+# Add a new asset class
+
+The business is growing and activities are extending to livestock and meat. 
+This is as simple as adding `AssetClass.LIVESTOCK_AND_MEAT` value and updating the visitor interface.
+
+```java
+public enum AssetClass {
+    METAL {
+        @Override
+        public <E> E accept(AssetClassVisitor<E> visitor) {
+            return visitor.visitMetal();
+        }
+    },
+    ENERGY {
+        @Override
+        public <E> E accept(AssetClassVisitor<E> visitor) {
+            return visitor.visitEnergy();
+        }
+    },
+    AGRICULTURAL {
+        @Override
+        public <E> E accept(AssetClassVisitor<E> visitor) {
+            return visitor.visitAgricultural();
+        }
+    },
+    // The new value
+    LIVESTOCK_AND_MEAT {
+        @Override
+        public <E> E accept(AssetClassVisitor<E> visitor) {
+            return visitor.visitLiveStockAndMeat();
+        }
+    };
+
+    public abstract <E> E accept(AssetClassVisitor<E> visitor);
+}
+```
+
+```java
+public interface AssetClassVisitor<T> {
+    T visitMetal();
+    T visitEnergy();
+    T visitAgricultural();
+    // The new method
+    T visitLiveStockAndMeat();
+}
+```
+
+After this, the code will light up like a Christmas Tree: it does not compile anymore. The compiler should be thanked for doing such a great job! All those highlighted errors show that some part of the code is not designed to handle the new value yet. Let's correct this by using an exception: the existing features are not available yet for livestock and meat.
+
+
+```java
+public AutomatedTradingStrategy getAutomatedTradingStrategy(AssetClass assetClass) {
+    return assetClass.accept(new AssetClassVisitor<AutomatedTradingStrategy>() {
+        @Override
+        public AutomatedTradingStrategy visitMetal() {
+            return new HedgingStrategy();
+        }
+
+        @Override
+        public AutomatedTradingStrategy visitEnergy() {
+            return new SwingTradingStrategy();
+        }
+
+        @Override
+        public AutomatedTradingStrategy visitAgricultural() {
+            return new DayTradingStrategy();
+        }
+
+        @Override
+        public AutomatedTradingStrategy visitLiveStockAndMeat() {
+            throw new AutomatedTradingNotSupported("Automated trading for Livestock and meat is not allowed.")
+        }
+    });
+}
+```
+ 
+# In a nutshell
+
+During one of my projects, the team was dealing with a large number of enums and there was business logic based upon their values. The visitor pattern was our shield against unpredicted edge cases. It became our standard way to deal with enums in the code base.
+
+Using this pattern is not necessary if your enums are purely descriptive. However, bringing the heavy artillery is definitively worth the extra cost. Breaking the coupling between enum's values and the business logic offers flexibility while the compiler provides an instantaneous feedback about potential oversight and edge cases.
```

Comment:
Breaking the coupling between an enum's values and the business logic offers flexibility, while the compiler provides an instantaneous feedback about potential oversight and edge cases.